# Problemas de prueba de hipótesis

In [6]:
import pandas as pd
import numpy as np

import scipy.stats as stats

import matplotlib.pyplot as plt
import seaborn as sns

## Ejercicio 1

Usted es un nutricionista que investiga dos tipos diferentes de dietas para ver si existe una diferencia significativa en la pérdida de peso después de un mes. Eliges dos grupos aleatorios de personas; un grupo sigue la primera dieta y el otro sigue la segunda. Al final del mes, se registra la pérdida de peso (en kg) de cada persona:

| Diet 1 | Diet 2 |
|:-------|:-------|
| 2.0 | 3.0 |
| 2.5 | 3.2 |
| 3.0 | 3.1 |
| 2.8 | 2.9 |
| 2.3 | 2.8 |
| 2.7 | 3.0 |
| 2.5 | 3.2 |

Con estos datos se busca responder a la siguiente pregunta: ¿Existe una diferencia significativa en la pérdida de peso promedio entre las personas que siguieron la primera dieta y las que siguieron la segunda dieta?

Para sacar conclusiones, siga los siguientes puntos:

-Enunciar la hipótesis: hipótesis nula y alternativa.
-Realizar la prueba para comprobar la hipótesis. Puede utilizar la prueba t de Student.
-Analizar las conclusiones.

1. Enunciar la hipótesis
- Hipótesis nula: No existe diferencia significativa entre las dos dietas en términos de pérdida de peso: H₀
- Alternativa: Si hay una diferencia significativa entre las dos dietas en términos de pérdida de peso: H₁

In [ ]:
# TODO
#Realizar la prueba para comprobar la hipótesis: utilizar la prueba t de Student: Permite comparar dos grupos o categorías numéricas con un tamaño muestral reducido. Este caso es el contraste de hipótesis para la media en el caso de que se desconozca la varianza con muestra reducida. Sin embargo, también se puede usar cuando se conoce.
import numpy as np
from scipy import stats

Diet1 = [2.0, 2.5, 3.0, 2.8, 2.3, 2.7, 2.5]
Diet2 =[3.0, 3.2, 3.1, 2.9, 2.8, 3.0, 3.2]

media= np.mean(Diet1), np.mean(Diet2) 
print("Media Diet1:", f"{media[0]:.2f}")
print("Media Diet2:", f"{media[1]:.2f}")

t_statistic, p_value = stats.ttest_ind(Diet1, Diet2) 
#t_statistic: La medida de la diferencia entre los grupos en términos de desviaciones estándar. Un valor más alto indica una mayor diferencia entre los grupos.
#p_value: La probabilidad de obtener esta diferencia por puro azar si H₀ fuera verdad.

print("T-statistic:", t_statistic)
print("P-value:", p_value)

Media Diet1: 2.54
Media Diet2: 3.03
T-statistic: -3.5383407969933938
P-value: 0.004083270191713912


3. Analizar las conclusiones

Dado que p_value es muchisimo menor que 0.05, llegamos a la conclusión de que si que existe diferencia significativa entre ambas dietas. 

## ANOVA

**ANOVA** (*Analysis of Variance - Análisis de variación*) Es una técnica estadística utilizada para comparar las medidas de dos o más grupos. La idea detrás de ANOVA es descomponer la variabilidad total de los datos en dos componentes: variabilidad entre grupos y variabilidad dentro de los grupos:

- **Variabilidad entre grupos**: Esta variabilidad se refiere a las diferencias entre las medias del grupo. Si esta variabilidad es considerablemente mayor que la variabilidad dentro del grupo, podría ser una indicación de que al menos una de las medias del grupo es diferente.
- **Variabilidad dentro del grupo**: Esta variabilidad se refiere a la dispersión de los datos dentro de cada grupo. Si todos los grupos tienen una variabilidad similar, entonces cualquier diferencia notable en las medias de los grupos podría considerarse significativa.

Las hipótesis en ANOVA suelen incluir:

- **Hipótesis nula** ($H₀$): Las medias de todos los grupos son iguales.
- **Hipótesis alternativa** ($H₁$): Al menos una de las medias del grupo es diferente.

Si el resultado de la prueba ANOVA es significativo (por ejemplo, un valor p inferior a un umbral como 0,05), esto sugiere que al menos la media de un grupo es diferente.

## Ejercicio 2

Un agricultor decide probar tres tipos diferentes de fertilizantes para determinar si alguno es superior en términos de producción de maíz. El agricultor planta maíz en 15 parcelas idénticas y utiliza los tres fertilizantes (5 parcelas para cada tipo). Al final de la temporada mide el rendimiento de maíz (en kg) de cada parcela, con el siguiente resultado:

| Fertilizante 1 | Fertilizante 2 | Fertilizante 3 |
|:-------------|:-------------|:-------------|
| 20 | 22 | 24 |
| 21 | 21 | 23 |
| 20 | 23 | 22 |
| 19 | 22 | 23 |
| 20 | 21 | 24 |

Con estos datos busca responder la siguiente pregunta: ¿Existe una diferencia significativa en el rendimiento promedio del maíz entre los tres tipos de fertilizantes?

Para ayudarte, sigue los siguientes puntos:

-Enunciar la hipótesis: hipótesis nula y alternativa.
-Realizar la prueba ANOVA.
-Analizar las conclusiones.
-Si un fertilizante es mejor que otro, ¿cómo podemos saberlo?

1. Enunciar la hipótesis
- Hipótesis nula: No existe diferencia significativa en el rendimiento promeio del maíz entre los tres tipos de fertilizante
- Alternativa: Si existe una diferenci significativa en el rendimiento promedio del maíz entre los tres tipos de fertiliante

In [7]:
# TODO
f_1 =[20, 21, 20, 19, 20]
f_2 = [22, 21, 23, 22, 21]
f_3 = [24, 23, 22, 23, 24]

#Prueba ANOVA
f_statistic, p_value = stats.f_oneway(f_1, f_2, f_3)
print("F-statistic:", f_statistic)
print("P-value:", p_value)

F-statistic: 20.31578947368421
P-value: 0.00014047824793190402


3. Analizar las conclusiones

Dado que p_value es muchisimo menor que 0.05, llegamos a la conclusión de que si existe diferencia significativa en el rendimiento promedio del maíz entre al menos uno de los tres fertilizantes.

4. Un fertilizante es mejor que otro?

Para poder saber si un fertilizante es mejor que otro, deberiamos realizar comparaciones múltiples adhoc: con una prueba tukey

In [10]:
!pip install statsmodels

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.0/10.0 MB 18.4 MB/s eta 0:00:000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 233.3/233.3 kB 13.3 MB/s eta 0:00:00

[notice] A new release of pip is available: 24.0 -> 26.0.1
[notice] To update, run: pip install --upgrade pip


In [11]:
from scipy import stats
from statsmodels.stats.multicomp import pairwise_tukeyhsd
import numpy as np
import pandas as pd

# Combinamos todos los datos en un array
datos = np.concatenate([f_1, f_2, f_3])
# Creamos etiquetas para saber a qué grupo pertenece cada valor
etiquetas = ['F1']*len(f_1) + ['F2']*len(f_2) + ['F3']*len(f_3)

# Prueba de Tukey
resultado = pairwise_tukeyhsd(datos, etiquetas, alpha=0.05)
print(resultado)

Multiple Comparison of Means - Tukey HSD, FWER=0.05
group1 group2 meandiff p-adj  lower  upper  reject
--------------------------------------------------
    F1     F2      1.8 0.0099 0.4572 3.1428   True
    F1     F3      3.2 0.0001 1.8572 4.5428   True
    F2     F3      1.4 0.0409 0.0572 2.7428   True
--------------------------------------------------


Los tres fertilizantes son bastante diferentes entre si. Con el análisis turkey, llegamos a la conclusión de que el "ranking" sería: F3>F2>F2. 

Por lo que el mejor fertilizante sería el Fertilizante 3, con una diferenci de 3.2 sobre el Fertilizante 1 y de 1.4 sobre el Fertilizante 2.